### PR TITLE
 Implementation AC/DC (moment buffer reset & finish training on decompression epoch)

### PR DESCRIPTION
**Pull request by dbogunowicz**
AC_DC implementation ready, now testing!

### CODE REVIEW

File: ac_dc_recipy.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+<!--
```
Comment (bfineran) [inline ac_dc_recipy.md:1]: nit: commited a testing file

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 2]
```diff
@@ -1762,3 +1763,167 @@ def _should_log(
...
+        pruning to.  Regex patterns must be specified with the prefix 're:'. __ALL__
+        will match to all parameters. __ALL_PRUNABLE__ will match to all ConvNd
+        and Linear layers' weights. If a sparsity to param mapping is defined by
+        final_sparsity, then params should be set to []
+    """
+
+    def __init__(
+        self,
+        # because method does not involve any interpolation
+        # compression sparsity is a single float.
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1798]: nit: let's move this comment to line 1806

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 3]
```diff
@@ -1762,3 +1763,167 @@ def _should_log(
...
+            end_epoch=self._finish_on_decompression(
+                start_epoch, end_epoch, update_frequency
+            ),
+            update_frequency=update_frequency,
+        )
+
+        self.validate()
+
+    @staticmethod
+    def _finish_on_decompression(start_epoch, end_epoch, update_frequency):
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1825]: nit: private methods should go below public

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 4]
```diff
@@ -1762,3 +1763,167 @@ def _should_log(
...
+        for decompressed_phase_epoch in reversed(decompression_phases):
+            if decompressed_phase_epoch:
+                # when encounter a decompressed_phase_epoch: break
+                break
+            else:
+                # otherwise, keep subtracting last compression epochs count from the 'end epoch'
+                end_epoch -= 1
+        return end_epoch
+
+    def _check_mask_update(
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1851]: nit: private methods should go below public

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 5]
```diff
@@ -1762,3 +1763,167 @@ def _should_log(
...
+                    self._momentum_buffer_empty = False
+
+            self._module_masks.set_param_masks_from_sparsity(self._applied_sparsity)
+            self._sparsity_applied = True
+
+        if self.end_pending(epoch, steps_per_epoch):
+            self._module_masks.pruning_end(self._leave_enabled)
+
+    @staticmethod
+    def _reset_momentum_buffer(optimizer):
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1897]: nit: private methods should go below public

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 6]
```diff
@@ -1762,3 +1763,167 @@ def _should_log(
...
+        start_epoch: float,
+        end_epoch: float,
+        update_frequency: float,
+        params: Union[str, List[str]],
+    ):
+
+        self._compression_sparsity = compression_sparsity
+        self._decompression_sparsity = 0.0  # this is implicitly assumed in paper.
+        self._is_phase_decompression = True
+        self._momentum_buffer_empty = True
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:1809]: let's add an option to turn this functionality off and add a `ModifierProp` to control it

**Review by bfineran — CHANGES_REQUESTED** [review]
great first pass. after looking at `_check_mask_update`, let's definitely hold off on landing until we can migrate to the refactor as the logic will be easier going through the new flow
additionally, take a look at the tests currently in `tests/sparseml/pytorch/optim/test_modifier_pruning.py` and see how we can add testing for the `ACDCPruningModifier` (there's some logic in there already with the `phased` functionality and we have a yaml serialization test for each modifier)

**Comment by dbogunowicz** [comment]
@bfineran just for completeness, will the refactor of pruning modifiers be compatible with the current unit tests? Just wondering whether it makes sense for me to write tests before we move to the new modifier structure.

**Comment by bfineran** [comment]
> @bfineran thank you for your comments, excited about my first, real pull request :) Just for completeness, will the refactor of pruning modifiers be compatible with the current unit tests? Just wondering whether it makes sense for me to write tests before we move to the new modifier structure.

the general test structure and classes will remain

File: src/sparseml/pytorch/optim/modifier_pruning.py (hunk 2 above)
Comment (dbogunowicz) [reply src/sparseml/pytorch/optim/modifier_pruning.py:1798]: Agree, makes sense.

File: src/sparseml/pytorch/optim/modifier_pruning.py [hunk 7]
```diff
@@ -578,6 +578,30 @@ def applied_sparsity(self) -> Union[float, List[float]]:
...
+    @ModifierProp(serializable=False)
+    def inter_func(self) -> float:
+        """
+        :return: the initial sparsity for the param to start with at
+        start_epoch
+        """
+        return self._inter_func
+
+    @ModifierProp(serializable=False)
+    def log_types(self) -> float:
```
Comment (bfineran) [inline src/sparseml/pytorch/optim/modifier_pruning.py:598]: we want to add log types to this modifier's constructor

File: ac_dc_recipy.md (hunk 1 above)
Comment (bfineran) [reply ac_dc_recipy.md:1]: @dbogunowicz bump to remove

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+        self,
+        compression_sparsity: float,
+        start_epoch: float,
+        end_epoch: float,
+        update_frequency: float,
+        params: Union[str, List[str]],
+        global_sparsity: bool = True,
+        leave_enabled: bool = True,
+        momentum_buffer_reset: bool = True,
+        mask_type: Union[str, List[int], PruningMaskCreator] = "unstructured",
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:94]:
```suggestion
        mask_type: str = "unstructured",
```

only supports strings now

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        :param epoch: current epoch and progress within the current epoch
+        :param steps_per_epoch: number of steps taken within each epoch
+            (calculate batch number using this and epoch)
+        """
+        super().optimizer_post_step(module, optimizer, epoch, steps_per_epoch)
+
+        # be sure to apply mask again after optimizer update because
+        # weights may have changed (optimizer with momentum, not masking gradient)
+
+        self._module_masks.apply()
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:137]: this is handled in the super class, right?

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+            This condition is only evaluated when `momentum_buffer_reset`
+            strategy is True. When entering decompression phase check
+            whether momentum buffer is empty. If not
+            (this happens always before the first epoch of the decompression phase),
+            reset!
+            """
+            self._reset_momentum_buffer(optimizer)
+            self._momentum_buffer_empty = True
+
+        self._sparsity_applied = False  # damian: is this necessary?
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:153]: also handled in super class

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        global_sparsity: bool = True,
+        leave_enabled: bool = True,
+        momentum_buffer_reset: bool = True,
+        mask_type: Union[str, List[int], PruningMaskCreator] = "unstructured",
+    ):
+
+        # because method does not involve any interpolation
+        # compression sparsity (final sparsity) is a single float.
+        self._compression_sparsity = compression_sparsity
+        self._decompression_sparsity = 0.0  # this is implicitly assumed in paper.
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:100]: nice, good to potentially have for later

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py (hunk 10 above)
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:153]: also handled in super class

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py [hunk 12]
```diff
@@ -117,10 +117,10 @@ def __init__(
         log_types: Union[str, List[str]] = None,
         global_sparsity: bool = False,
         allow_reintroduction: bool = False,
+        leave_enabled: bool = False,
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:120]: noting that these changes will be replaced by multi-inheretance patch discussed earlier today

**Review by bfineran — CHANGES_REQUESTED** [review]
refactored implementation looks really clean. looks good from my end after comments are resolved and tests added

**Review by bfineran — DISMISSED** [review]

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py (hunk 12 above)
Comment (bfineran) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:128]: @dbogunowicz bumping this one - is this still necessary / does anything break by adding this now?

Comment (dbogunowicz) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:128]: `leave_enabled` works correctly, as expected, and is compatible with Konstantin's changes as well as all the tests :)

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+            reset!
+            """
+            self._reset_momentum_buffer(optimizer)
+            self._momentum_buffer_empty = True
+
+    def get_applied_sparsity_for_epoch(
+        self, epoch: float, steps_per_epoch: int
+    ) -> Union[float, List[float]]:
+        """
+        :param epoch: curent epoch
```
Comment (KSGulin) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py:162]: nit: 'current' typy

**Review by KSGulin — COMMENTED** [review]
Looks great! Only found one small nit.

**Review by KSGulin — DISMISSED** [review]

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py (hunk 12 above)
Comment (dbogunowicz) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:128]: ...and also necessary imo:
- setting `leave_enabled=False` functionality is helpful to reproduce the original paper, where after alternating between compression/decompression phases with `update_frequency=5`, we finish off the training with `10 decompression` and finally `15 compression` fine-tuning epochs. So in that case, it is useful to stop masking the weights after last epoch of the modifier, just before the 10-epoch fine-tuning decompression phase.
- setting `leave_enabled=True` would be desirable by most of the users to simply use this modifier out-of-the-box and retrieve the sparse model.

Also, in the paper Dan's students experiment with having dense-only/sparse-only fine tuning after the compression/decompression phases, so all in all, I like having this argument there.

Having said that, if you agree with my arguments, I should probably make this argument `True` by default (to make it easier to use in out-of-the-box fashion).

**Review by bfineran — CHANGES_REQUESTED** [review]

File: tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_acdc.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+        yaml_modifier.start_epoch
+        == serialized_modifier.start_epoch
+        == obj_modifier.start_epoch
+    )
+    assert (
+        yaml_modifier.end_epoch
+        == serialized_modifier.end_epoch
+        == obj_modifier.end_epoch
+    )
+    assert yaml_modifier.params == serialized_modifier.params == obj_modifier.params
```
Comment (bfineran) [inline tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_acdc.py:255]: need to add tests for compression sparsity, leave_enabled and others

**Review by bfineran — CHANGES_REQUESTED** [review]

File: tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_acdc.py (hunk 14 above)
Comment (dbogunowicz) [reply tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_acdc.py:255]: You mean basically a group of tests to make sure that those functionalities work properly right?

**Comment by bfineran** [comment]
serialization fixes added

**Review by bfineran — DISMISSED** [review]

**Review by bfineran — APPROVED** [review]

**Review by markurtz — APPROVED** [review]
